**Goodbooks Recommendation Engine**

Yiyan Zhang

**1. Business Understanding**

The purpose of this project is to build a simple recommendation engine to recommend books based on reader's ratings.

**2. Data Understanding**

The dataset is from [Kaggle](https://www.kaggle.com/zygmunt/goodbooks-10k).

The dataset contains the following files:
* ratings.csv: Users' ratings for books that they have read.
* books.csv: Metadata for each book.
* to_read.csv: Book marked "to read" by users, like a reading wish list.
* book_tags.csv: User assigned tags/generes to books.
* tags.csv: Translatre tag IDs to names.

**Data Importing**

In [5]:
# import ratings data
ratings = spark.read.csv('/FileStore/tables/ratings.csv',
                        header=True,
                        sep=',',
                        inferSchema=True)
ratings.printSchema()

root
-- book_id: integer (nullable = true)
-- user_id: integer (nullable = true)
-- rating: integer (nullable = true)

In [6]:
# import to_read data
to_read = spark.read.csv('/FileStore/tables/to_read.csv',
                        header=True,
                        sep=',',
                        inferSchema=True)
to_read.printSchema()

root
-- user_id: integer (nullable = true)
-- book_id: integer (nullable = true)

In [7]:
#import books data
books = spark.read.csv('/FileStore/tables/books.csv',
                        header=True,
                        sep=',',
                        inferSchema=True)
books.printSchema()

root
-- id: integer (nullable = true)
-- book_id: integer (nullable = true)
-- best_book_id: integer (nullable = true)
-- work_id: integer (nullable = true)
-- books_count: integer (nullable = true)
-- isbn: string (nullable = true)
-- isbn13: double (nullable = true)
-- authors: string (nullable = true)
-- original_publication_year: double (nullable = true)
-- original_title: string (nullable = true)
-- title: string (nullable = true)
-- language_code: string (nullable = true)
-- average_rating: string (nullable = true)
-- ratings_count: string (nullable = true)
-- work_ratings_count: string (nullable = true)
-- work_text_reviews_count: string (nullable = true)
-- ratings_1: double (nullable = true)
-- ratings_2: integer (nullable = true)
-- ratings_3: integer (nullable = true)
-- ratings_4: integer (nullable = true)
-- ratings_5: integer (nullable = true)
-- image_url: string (nullable = true)
-- small_image_url: string (nullable = true)

The "original_title" column in books.csv has strings with commas in them, when importing using comma as delimiter, some entries in the resulted tables were misaligned to the column header.

I tried to format the CSV file and put quotation marks around titles, but couldn't solve the problem. I also tried saving the file as a TSV file and used tab as delimiter, but still couldn't achieve what I wanted.

The quotation marks existed after uploding the CSV file to Databricks FileStore but were missing after importing into the cluster attached to the notebook.

However this problem did not affect the modelling process.

In [9]:
# import book_tags data
book_tags = spark.read.csv('/FileStore/tables/book_tags.csv',
                        header=True,
                        sep=',',
                        inferSchema=True)
book_tags.printSchema()

root
-- goodreads_book_id: integer (nullable = true)
-- tag_id: integer (nullable = true)
-- count: integer (nullable = true)

In [10]:
#import tags data
tags = spark.read.csv('/FileStore/tables/tags.csv',
                        header=True,
                        sep=',',
                        inferSchema=True)
tags.printSchema()

root
-- tag_id: integer (nullable = true)
-- tag_name: string (nullable = true)

Once data was imported into Databricks, I conducted exploratory data analysis (EDA) on the datasets to get familiar with the data, to identify data sanity issues, and to discover first-hand insights.

**EDA on ratings**

In [13]:
import pandas as pd

# convert ratings to pandas dataframe
ratings_df = ratings.toPandas()
ratings_df.head()

book_id 
 user_id 
 rating 
 
 
 
 
 0 
 1 
 314 
 5 
 
 
 1 
 1 
 439 
 3 
 
 
 2 
 1 
 588 
 5 
 
 
 3 
 1 
 1169 
 4 
 
 
 4 
 1 
 1185 
 4

In [14]:
# check uniques and missing values
total = ratings_df['rating'].count()
b_num = ratings_df['book_id'].nunique()
u_num = ratings_df['user_id'].nunique()
na = ratings_df.isnull().values.any()

# average ratings
b_rating = round((u_num/b_num), 2)
avg_rating = round((total/b_num), 2)
u_rating = round((total/u_num), 2)

# missing ratings
missed_rating = (1 - (total/(b_num * u_num))) * 100

print('There are', b_num, 'distinct books and', u_num, 'individual users.\n', 
      'Are there missing values?:', na, '\n', 
      'On average, one book is rated by', b_rating, 
      'distinct users and has', avg_rating, 'ratings.\n',
      'One user gives out', u_rating, 'ratings.\n',
      'Not all users rate all books, so', '%.2f' % missed_rating, '% of the ratings is missing.')

There are 10000 distinct books and 53424 individual users.
 Are there missing values?: False 
 On average, one book is rated by 5.34 distinct users and has 98.18 ratings.
 One user gives out 18.38 ratings.
 Not all users rate all books, so 99.82 % of the ratings is missing.

In [15]:
# check ratings with histogram to see distribution
ratings_df.hist(column='rating', bins=5)

In [16]:
# get summary of ratings
ratings_df['rating'].describe()

Out[17]: count 981756.000000
mean 3.856534
std 0.983941
min 1.000000
25% 3.000000
50% 4.000000
75% 5.000000
max 5.000000
Name: rating, dtype: float64

Ratings range from 1 to 5 with an average of 3.86 and a standard deviation of 0.98.
There is no outlier in the data and majority of the ratings are high.

**EDA on to_read**

In [19]:
# convert to_read to pandas dataframe
read_df = to_read.toPandas()
read_df.head()

user_id 
 book_id 
 
 
 
 
 0 
 1 
 112 
 
 
 1 
 1 
 235 
 
 
 2 
 1 
 533 
 
 
 3 
 1 
 1198 
 
 
 4 
 1 
 1874

In [20]:
# check uniques
rows = len(read_df)
b_num = read_df['book_id'].nunique()
u_num = read_df['user_id'].nunique()
avg_read = round((rows/b_num),2)

print('On average, one user has', avg_read, 'book in the to read list.')

On average, one user has 91.4 book in the to read list.

In [21]:
from pyspark.sql import Row

# create temporary tables
to_read.createOrReplaceTempView('to_read')
books.createOrReplaceTempView('books')

In [22]:
%sql
-- top 10 users with the most number of books in their to read list --
SELECT user_id, COUNT(*) as Num
FROM to_read
GROUP BY user_id
ORDER BY Num DESC LIMIT 10;

user_id Num 38457 117 28259 114 38076 110 44530 108 46555 107 46000 107 34162 105 34487 104 39174 103 40362 102

In [23]:
%sql
-- top 10 saved books in users' to read list --
SELECT a.book_id, b.title, COUNT(*) as Num
FROM to_read a
INNER JOIN books b
ON a.book_id = b.book_id
GROUP BY a.book_id, b.title
ORDER BY Num DESC LIMIT 10;

book_id title Num 13 The Ultimate Hitchhiker's Guide to the Galaxy 1812 11 The Hitchhiker's Guide to the Galaxy (Hitchhiker's Guide to the Galaxy, #1) 1767 119 The Lord of the Rings: The Art of The Fellowship of the Ring 1499 8 Harry Potter Boxed Set, Books 1-5 (Harry Potter, #1-5) 1498 6 Harry Potter and the Goblet of Fire (Harry Potter, #4) 1484 67 The Known World 1352 5 Harry Potter and the Prisoner of Azkaban (Harry Potter, #3) 1293 36 The Lord of the Rings: Weapons and Warfare 1211 28 Notes from a Small Island 1148 10 Harry Potter Collection (Harry Potter, #1-6) 1110

**EDA on books**

In [25]:
# convert to pandas dataframe
books_df = books.toPandas()
books_df.head()

# remove entries due to data importing issue
books_df = books_df[books_df.ratings_count.str.isdigit() == True]
books_df.ratings_count = books_df.ratings_count.astype('int')
books_df.average_rating = books_df.average_rating.astype('float')

I originally wanted to use SQL to generate the tables below, but I could not solve the data importing issue. Instead I utilized Pandas and manipulated the dataframe to get the resulted tables.

In [27]:
# top 10 popular books' ratings
books_df.sort_values(by = 'ratings_count', ascending = False)[['original_title',
                                                               'ratings_count', 
                                                               'average_rating' ]][0:10]

original_title 
 ratings_count 
 average_rating 
 
 
 
 
 0 
 The Hunger Games 
 4780653 
 4.34 
 
 
 1 
 Harry Potter and the Philosopher's Stone 
 4602479 
 4.44 
 
 
 2 
 Twilight 
 3866839 
 3.57 
 
 
 3 
 To Kill a Mockingbird 
 3198671 
 4.25 
 
 
 4 
 The Great Gatsby 
 2683664 
 3.89 
 
 
 5 
 The Fault in Our Stars 
 2346404 
 4.26 
 
 
 6 
 The Hobbit or There and Back Again 
 2071616 
 4.25 
 
 
 7 
 The Catcher in the Rye 
 2044241 
 3.79 
 
 
 9 
 Pride and Prejudice 
 2035490 
 4.24 
 
 
 8 
 Angels & Demons 
 2001311 
 3.85

In [28]:
# top 10 rated books
books_df.sort_values(by = 'average_rating', ascending = False)[['original_title',
                                                                'average_rating' ]][0:10]

original_title 
 average_rating 
 
 
 
 
 3627 
 The Complete Calvin and Hobbes 
 4.82 
 
 
 861 
 Words of Radiance 
 4.77 
 
 
 3274 
 None 
 4.77 
 
 
 8853 
 Mark of the Lion Trilogy 
 4.76 
 
 
 7946 
 None 
 4.76 
 
 
 4482 
 It's a Magical World: A Calvin and Hobbes Coll... 
 4.75 
 
 
 6360 
 There's Treasure Everywhere: A Calvin and Hobb... 
 4.74 
 
 
 421 
 Complete Harry Potter Boxed Set 
 4.74 
 
 
 6589 
 The Authoritative Calvin and Hobbes 
 4.73 
 
 
 6919 
 The Indispensable Calvin and Hobbes: A Calvin ... 
 4.73

**EDA on book_tags and tags**

In [30]:
# EDA on book_tags
book_tags.show(n=5)

+-----------------+------+------+
goodreads_book_id|tag_id| count|
+-----------------+------+------+
 1| 30574|167697|
 1| 11305| 37174|
 1| 11557| 34173|
 1| 8717| 12986|
 1| 33114| 12716|
+-----------------+------+------+
only showing top 5 rows

In [31]:
# EDA on tags
tags.show(n=5)

+------+--------+
tag_id|tag_name|
+------+--------+
 0| -|
 1| --1-|
 2| --10-|
 3| --12-|
 4| --122-|
+------+--------+
only showing top 5 rows

In [32]:
# create temporary tables
book_tags.createOrReplaceTempView('book_tags')
tags.createOrReplaceTempView('tags')

In [33]:
%sql
-- 10 most popular tag names --
SELECT DISTINCT a.tag_name, SUM(b.count) as Total
FROM tags a
INNER JOIN book_tags b
ON a.tag_id = b.tag_id
GROUP BY a.tag_name
ORDER BY Total DESC LIMIT 10;

tag_name Total to-read 140718761 currently-reading 7507958 favorites 4503173 fiction 3688819 fantasy 3548157 young-adult 1848306 classics 1756920 books-i-own 1317235 romance 1231926 owned 1224279

**3. Data Preparation**

The raw datasets is in healthy shape and is ready to start modeling.

**4. Modeling**

I utilized collaborative filtering and used Alternating Least Square (ALS) to build the recommendation engine. 

Collaborative filtering is an important and popular concept behind recommendation engine, it basicly means that people with similar taste tend to consume the same kind of product. 

Collaborative filtering can be achieved by matrix factorization, where a user-item matrix can be factored into a user-feature matrix and an item-feature matrix. While ALS is used to find the latent factor, the link between users and their ratings, and to approximate the factor weights to minimize the least squares between prediction and actual ratings. As a result, ALS will fill in missing ratings with its predictions.

I set coldStartStrategy to 'drop' because I want to make sure that there is no NAs in the predictions and the evaluator will return meaningful results. ImplicitPrefs is set to False because the data has users' ratings and ALS will recognize them as explicit preferences. NonNegative is just to make sure there will be no negative numbers in the predictions.

In [36]:
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.recommendation import ALS
from pyspark.ml.tuning import TrainValidationSplit, ParamGridBuilder, CrossValidator 

# split data into training and test set
(train, test) = ratings.randomSplit([0.8, 0.2])

# create ALS model
als = ALS(userCol='user_id', itemCol='book_id', ratingCol='rating',
          coldStartStrategy='drop', nonnegative=True, implicitPrefs=False)

# define evaluator as RMSE
evaluator = RegressionEvaluator(metricName = 'rmse', 
                                labelCol = 'rating', 
                                predictionCol = 'prediction')

In [37]:
# tune model
param_grid = ParamGridBuilder() \
            .addGrid(als.rank, [10, 20, 75, 150]) \
            .addGrid(als.maxIter, [10, 50, 100]) \
            .addGrid(als.regParam, [0.01, 0.05, 0.1]) \
            .build()

# build cross validator
cv = CrossValidator(estimator = als, 
                    estimatorParamMaps = param_grid, 
                    evaluator = evaluator, 
                    numFolds = 5)

In [38]:
# fit model
model = als.fit(train)

# make predictions
pred = model.transform(test)

In [39]:
# create function to tune hyper parameter and find the best model configuration
def tune_als(train_data, test_data, ranks, iters, params):
  """
  train_data: dataset for training
  test_data: dataset for testing
  ranks: number of latent factors
  iters: MaxIter
  params: RegParam
  return: the best fitted ALS model with the lowest RMSE
  """
  
  # initialize place holders
  min_error = float('inf')
  best_rank = -1
  best_iter = 1
  best_reg = 0
  best_model = None
  
  for rank in ranks:
    for reg in params:
      
      # configure ALS model
      als = ALS()\
      .setRank(rank)\
      .setMaxIter(iters)\
      .setRegParam(reg)\
      .setUserCol("user_id")\
      .setItemCol("book_id")\
      .setRatingCol("rating")
      
      # train model
      model = als.fit(train_data)
      
      # predict and evaluate model
      pred = model.transform(test_data)
      evaluator = RegressionEvaluator(metricName = 'rmse', 
                              labelCol = 'rating', 
                              predictionCol = 'prediction')
      rmse = evaluator.evaluate(pred)

      if rmse < min_error:
        min_error = rmse
        best_rank = rank
        best_reg = reg
        best_model = model
  print('Rank:', best_rank)
  print('Iter:', best_iter)
  print('Param:', best_reg)
  return best_model


# hyper param config
iters = 50
ranks = [10, 50, 100]
reg_param = [0.01, 0.05, 0.1]

final_model = tune_als(train, test, ranks, iters, reg_param)

**5. Evaluation**

I wanted to find the best hyper parameters for the ALS model and tried cross validation and user defined function, but the process took extremely long time to complete and Databricks automatically logged me out during the session. I think the reason might be that I am running the free community version.

Instead I manually tested several groups of hyper parameters and found that the default setting has a RMSE of around 0.91, which I believe is acceptable and reasonable given the context that the engine is recommending books to readers. This number indicates that the model's predictions on ratings are 0.91 unit away from the actual ratings.

In [42]:
# calculate RMSE
rmse = evaluator.evaluate(pred)
print('RMSE is ' + str(rmse))

RMSE is 0.914249754717826

In [43]:
# create temporary tables
pred.createOrReplaceTempView('pred')
books.createOrReplaceTempView('books')

In [44]:
%sql
-- check model predictions --
SELECT pred.user_id, pred.book_id, pred.rating, pred.prediction, books.title 
FROM pred 
LEFT JOIN books 
ON pred.book_id = books.id

user_id book_id rating prediction title 588 148 4 3.365916 Girl with a Pearl Earring 9731 148 3 3.031473 Girl with a Pearl Earring 10727 148 4 3.300964 Girl with a Pearl Earring 9165 148 3 3.758632 Girl with a Pearl Earring 32055 148 3 3.12183 Girl with a Pearl Earring 10140 148 3 2.8816402 Girl with a Pearl Earring 7001 148 4 3.8797023 Girl with a Pearl Earring 20967 148 3 3.8291576 Girl with a Pearl Earring 10610 148 4 3.8230343 Girl with a Pearl Earring 14603 148 4 3.6024997 Girl with a Pearl Earring 33065 148 3 3.6438575 Girl with a Pearl Earring 18361 148 4 3.7142644 Girl with a Pearl Earring 29703 148 4 3.707699 Girl with a Pearl Earring 11569 148 2 3.5178268 Girl with a Pearl Earring 29031 148 3 3.4149473 Girl with a Pearl Earring 40167 148 5 3.3457234 Girl with a Pearl Earring 45554 148 4 3.6223578 Girl with a Pearl Earring 2171 148 4 3.4242601 Girl with a Pearl Earring 11239 148 2 2.6012306 Girl with a Pearl Earring 51166 148 4 3.8552058 Girl with a Pearl Earring 25840 148 3 3.5822127 Girl with a Pearl Earring 32748 148 4 3.9200835 Girl with a Pearl Earring 39618 148 3 3.3138366 Girl with a Pearl Earring 44397 148 5 4.1270447 Girl with a Pearl Earring 18031 148 5 4.013987 Girl with a Pearl Earring 11945 148 4 4.044533 Girl with a Pearl Earring 6689 463 4 3.1922424 Dragonfly in Amber (Outlander, #2) 17207 463 4 3.838755 Dragonfly in Amber (Outlander, #2) 24326 463 4 3.6253953 Dragonfly in Amber (Outlander, #2) 48111 463 3 2.7937865 Dragonfly in Amber (Outlander, #2) 29732 463 5 4.747379 Dragonfly in Amber (Outlander, #2) 23291 463 5 5.00362 Dragonfly in Amber (Outlander, #2) 11073 463 5 4.219201 Dragonfly in Amber (Outlander, #2) 40167 463 5 3.6444108 Dragonfly in Amber (Outlander, #2) 25164 463 3 3.2198484 Dragonfly in Amber (Outlander, #2) 6895 463 4 4.2989683 Dragonfly in Amber (Outlander, #2) 5272 463 5 3.4994092 Dragonfly in Amber (Outlander, #2) 2166 463 2 3.2796385 Dragonfly in Amber (Outlander, #2) 52740 463 4 3.1391077 Dragonfly in Amber (Outlander, #2) 20782 463 3 4.1294107 Dragonfly in Amber (Outlander, #2) 19942 471 3 3.8707917 Mansfield Park 11868 471 5 3.5286753 Mansfield Park 32419 471 4 3.7730136 Mansfield Park 725 471 5 3.5765605 Mansfield Park 4536 471 3 3.427395 Mansfield Park 14385 471 4 3.8533516 Mansfield Park 31996 471 3 3.3428326 Mansfield Park 30184 471 5 5.0007772 Mansfield Park 9806 471 4 3.471784 Mansfield Park 26718 471 4 3.5356321 Mansfield Park 23497 471 5 4.2673206 Mansfield Park 23004 471 5 4.65703 Mansfield Park 34363 471 3 3.057114 Mansfield Park 17329 471 4 3.3711214 Mansfield Park 8167 471 2 3.2133205 Mansfield Park 34688 471 4 3.2119026 Mansfield Park 7152 496 4 3.206971 The Clan of the Cave Bear (Earth's Children, #1) 11445 496 4 3.7843208 The Clan of the Cave Bear (Earth's Children, #1) 14248 496 5 3.8159218 The Clan of the Cave Bear (Earth's Children, #1) 11569 496 3 3.807827 The Clan of the Cave Bear (Earth's Children, #1) 29031 496 3 3.7398531 The Clan of the Cave Bear (Earth's Children, #1) 16377 496 5 4.245586 The Clan of the Cave Bear (Earth's Children, #1) 45554 496 3 3.7099888 The Clan of the Cave Bear (Earth's Children, #1) 9125 496 1 2.0837696 The Clan of the Cave Bear (Earth's Children, #1) 13112 496 3 3.6855862 The Clan of the Cave Bear (Earth's Children, #1) 11692 496 5 3.8980503 The Clan of the Cave Bear (Earth's Children, #1) 9804 496 5 3.5106866 The Clan of the Cave Bear (Earth's Children, #1) 33872 496 5 3.7662218 The Clan of the Cave Bear (Earth's Children, #1) 10335 496 3 3.6080713 The Clan of the Cave Bear (Earth's Children, #1) 25278 496 5 4.162655 The Clan of the Cave Bear (Earth's Children, #1) 10944 496 4 4.003278 The Clan of the Cave Bear (Earth's Children, #1) 28343 833 3 3.3517218 The Chamber 26992 833 3 3.198441 The Chamber 19942 833 3 3.3014128 The Chamber 17984 833 3 3.6470084 The Chamber 21268 833 3 3.6381016 The Chamber 2619 833 2 3.2307725 The Chamber 31615 833 3 3.7222629 The Chamber 23498 833 4 3.575051 The Chamber 4490 833 3 3.3907835 The Chamb

**Generate recommendations**

In [46]:
# generate 10 recommendations for all users
rec = model.recommendForAllUsers(10)
rec.show()

+-------+--------------------+
user_id| recommendations|
+-------+--------------------+
 148|[[9842, 4.6074276...|
 463|[[8638, 4.9583983...|
 471|[[2236, 3.043882]...|
 496|[[562, 5.753499],...|
 833|[[3124, 5.5570407...|
 1088|[[9024, 4.7818785...|
 1238|[[8571, 4.583519]...|
 1342|[[9531, 4.6892643...|
 1580|[[9842, 5.0372953...|
 1591|[[7844, 5.140592]...|
 1645|[[5990, 4.4841413...|
 1829|[[9842, 4.7628326...|
 1959|[[7844, 3.5534396...|
 2122|[[8362, 4.6946716...|
 2142|[[5376, 4.9902253...|
 2366|[[4868, 6.0180516...|
 2659|[[4868, 4.647514]...|
 2866|[[9842, 4.8812513...|
 3175|[[2840, 4.982255]...|
 3749|[[8249, 4.0800376...|
+-------+--------------------+
only showing top 20 rows

In [47]:
# create temporary tables
rec.createOrReplaceTempView('rec')

# transform data frame into cleaner format
tidy_rec = spark.sql(
  """
  SELECT user_id,
  x.book_id AS book_id,
  x.rating AS prediction
  FROM rec
  LATERAL VIEW explode(recommendations) exploded_table AS x
  """)

# create temporary tables
tidy_rec.createOrReplaceTempView('tidy_rec')

In [48]:
%sql
-- select the final recommendations --
SELECT a.user_id, a.book_id, a.prediction, b.title
FROM tidy_rec a
LEFT JOIN books b
ON a.book_id = b.book_id
WHERE title <> 'null'
ORDER BY a.user_id, a.prediction DESC;

user_id book_id prediction title 1 4264 3.7688138 Fever Pitch 1 5344 3.7579236 Hard Times 2 5344 5.3379664 Hard Times 3 5872 1.0544994 Regeneration (Regeneration, #1) 5 3885 5.175308 The Taste of Home Cookbook 7 5346 4.8078547 The Last Juror 8 9531 5.530328 Peter and the Shadow Thieves (Peter and the Starcatchers, #2) 11 5344 4.734245 Hard Times 12 4264 4.2636156 Fever Pitch 12 7455 4.123007 Sex and the City 14 6862 4.0720406 Amsterdam 14 9566 4.0488434 Still Life with Woodpecker 15 4264 3.9011798 Fever Pitch 15 6820 3.8383975 number9dream 16 6425 4.2414064 Hard Eight (Stephanie Plum, #8) 16 9516 4.069159 Persepolis: The Story of a Childhood (Persepolis, #1) 17 9570 4.320135 Another Roadside Attraction 18 5344 4.541117 Hard Times 18 4264 4.5096483 Fever Pitch 18 9531 4.5039577 Peter and the Shadow Thieves (Peter and the Starcatchers, #2) 18 6862 4.445798 Amsterdam 19 7661 3.9154062 Next 20 9531 4.886122 Peter and the Shadow Thieves (Peter and the Starcatchers, #2) 21 6425 4.183122 Hard Eight (Stephanie Plum, #8) 22 9531 4.4164047 Peter and the Shadow Thieves (Peter and the Starcatchers, #2) 24 3885 4.341229 The Taste of Home Cookbook 24 6862 4.2256765 Amsterdam 25 6862 5.261019 Amsterdam 27 5344 4.8745503 Hard Times 28 9531 4.6998596 Peter and the Shadow Thieves (Peter and the Starcatchers, #2) 28 6862 4.466911 Amsterdam 29 6194 5.0034876 Waiting for the Barbarians 30 5344 3.9658413 Hard Times 32 2526 4.834288 Blindness 34 9570 3.106571 Another Roadside Attraction 35 9566 3.955457 Still Life with Woodpecker 35 6862 3.8317146 Amsterdam 36 7039 4.9084563 Tears of the Giraffe (No. 1 Ladies' Detective Agency, #2) 36 4004 4.9063635 The Pilgrimage 37 5344 5.206152 Hard Times 39 6862 5.3942294 Amsterdam 41 3885 5.1024027 The Taste of Home Cookbook 42 4264 5.018747 Fever Pitch 42 5344 5.002961 Hard Times 43 9531 4.328982 Peter and the Shadow Thieves (Peter and the Starcatchers, #2) 45 2068 3.815545 Darwin's Dangerous Idea: Evolution and the Meanings of Life 46 4591 5.3440294 Genome: the Autobiography of a Species in 23 Chapters 47 25 4.830361 I'm a Stranger Here Myself: Notes on Returning to America after Twenty Years Away 49 6862 4.0887403 Amsterdam 51 4599 5.018753 Carter Beats the Devil 51 9569 4.9728446 Villa Incognito 51 6425 4.9350567 Hard Eight (Stephanie Plum, #8) 51 5453 4.919846 Forever in Blue: The Fourth Summer of the Sisterhood (Sisterhood, #4) 52 3885 5.013913 The Taste of Home Cookbook 54 4980 4.5146537 Breakfast of Champions 55 5454 4.445919 The Second Summer of the Sisterhood (Sisterhood, #2) 58 6862 4.746535 Amsterdam 58 5344 4.7324657 Hard Times 59 6613 4.5398054 Four Blondes 59 3885 4.4798636 The Taste of Home Cookbook 60 5344 4.915615 Hard Times 61 9531 4.739396 Peter and the Shadow Thieves (Peter and the Starcatchers, #2) 62 7081 4.847894 The Android's Dream 63 9598 4.6181173 Deadeye Dick 64 5849 4.543983 A House for Mr Biswas 66 9531 4.6171184 Peter and the Shadow Thieves (Peter and the Starcatchers, #2) 67 3692 4.829307 The Heart of the Matter 70 9534 2.8272789 Shadow of the Hegemon (Ender's Shadow, #2) 71 9531 3.7858894 Peter and the Shadow Thieves (Peter and the Starcatchers, #2) 72 7039 4.9198995 Tears of the Giraffe (No. 1 Ladies' Detective Agency, #2) 72 5344 4.900735 Hard Times 72 4264 4.8788595 Fever Pitch 74 3885 3.2745671 The Taste of Home Cookbook 77 2068 2.8616588 Darwin's Dangerous Idea: Evolution and the Meanings of Life 80 5344 3.059567 Hard Times 81 8492 2.071095 Prophet 82 5849 4.9702344 A House for Mr Biswas 82 9531 4.924454 Peter and the Shadow Thieves (Peter and the Starcatchers, #2) 84 9531 4.9289017 Peter and the Shadow Thieves (Peter and the Starcatchers, #2) 84 8492 4.640177 Prophet 86 6613 4.824363 Four Blondes 86 7039 4.7967415 Tears of the Giraffe (No. 1 Ladies' Detective Agency, #2) 88 3692 4.9789867 The Heart of the Matter 89 9566 4.552145 Still Life with Woodpecker 91 2068 4.2764654 Darwin's Dangerous Idea: Evolution and the Meanings of Life 91 9784 4.17142 Women in Love (Brangwen F

In [49]:
%sql
-- predicted ratings for to read list --
SELECT a.user_id, a.book_id, b.prediction AS rating_pred
FROM to_read a
INNER JOIN tidy_rec b
ON a.user_id = b.user_id AND a.book_id = b.book_id;

user_id book_id rating_pred 7833 862 4.504927 21715 8563 4.0605936 51454 8362 5.307687 15173 4104 4.103098 19619 6582 5.0904326 20052 2290 3.9824455 25106 3459 4.981147 52100 3093 4.691145 3876 9830 5.014241 13898 964 4.0606694 14809 987 3.8250403 14809 2441 3.7338557 20624 1275 5.110715 22185 94 5.0791035 38910 8362 5.5201807 38910 9842 5.512821 43307 562 4.756651 51455 192 3.8711426 4294 3836 3.8361847 31646 2840 5.350981 35514 1909 4.8578134 35756 1308 4.9088297 42589 8362 4.668378 48482 862 3.8834589 26746 7187 3.9877717 32061 4 4.2944436 51214 4032 4.1321135 6376 3952 4.233241 11287 862 4.432597 11402 5125 4.1130986 32860 5578 4.1137114 35782 1106 5.025965 40032 3628 5.260769 42545 6089 3.6807437 49318 2681 4.2365317 4697 562 3.956547 11894 39 5.206673 14760 9842 4.8947334 21909 3093 4.5678577 22966 3849 4.755709 42270 3435 5.3609056 3149 1597 4.3420844 3149 2290 4.265636 4259 4 4.519925 5222 562 4.5826645 12401 1385 4.710315 19678 6400 5.8949256 19532 6425 4.88219 19224 1198 3.9195426 52550 9842 5.139731 9209 2149 5.8059053 21956 2263 5.3833423 26529 1331 4.278271 29274 1577 4.1014457 34866 844 1.9051809 41281 3952 5.653505 52841 3093 4.0467954 1378 2256 4.226772 8442 143 4.249563 10180 5054 5.0317783 24371 2840 4.09526 36447 6591 4.473353 50284 1807 4.830338 14402 94 3.1007903 18742 2840 4.778562 34707 1808 4.0399413 183 983 4.042278 7522 1577 4.8816304 28374 7550 5.2936654 43774 1577 5.5855966 45828 3952 4.2495065 46087 6900 3.923081 51003 1577 4.6884995 53334 1264 4.0344706 11124 562 4.932133 26687 610 4.9153314 43590 2292 3.7488127 43590 3093 3.5349138 50458 192 4.217057 3151 4643 4.552726 8915 862 5.092349 34819 9973 4.7700105 49979 9252 5.606718 11883 5753 4.8541203 22918 1198 5.3907204 31322 7275 4.584655 36573 9946 5.3391633 36992 5604 5.0158553 45865 4 4.9288306 7351 3491 5.2653546 13081 3746 5.002717 14274 779 4.6865687 14720 8362 4.745589 19665 3952 4.3610535 27115 3885 4.612651 29868 5990 4.0089254 32955 6971 5.54691 33239 5933 4.5586 36380 8362 5.16138 36380 1577 5.135094 48587 7401 5.2477326 50978 2681 4.3896265 51600 5712 3.8869686 8962 3920 4.3737326 9120 1316 4.8796635 17895 6457 5.1909356 18819 7937 4.5783525 28716 8818 3.1244364 32831 389 5.2457986 6391 2844 2.9261875 8815 6102 4.5340476 14345 6287 4.3021855 37646 7831 5.3963666 5892 7424 4.437234 11240 4 1.040677 12067 8353 3.8609898 19010 1198 4.963881 52224 3248 4.8720303 20155 1499 3.7007816 23871 25 4.9779897 36716 1198 2.2894845 42068 2380 4.5271316 1260 2840 4.298298 17685 562 5.518796 20042 4032 5.107308 35219 7831 5.359731 52188 1385 3.673204 11151 8187 4.662274 11717 562 5.3964314 21726 5990 4.1520715 22519 7937 4.565434 40699 1499 5.021901 50779 3093 4.926681 5062 1810 4.357365 19212 2914 4.1878986 21246 2627 4.8051634 21246 3953 4.729634 29549 9531 4.970976 52191 4032 4.249957 4653 641 5.150746 21417 3124 5.202277 35341 143 3.7708085 41349 9405 3.8730757 42541 4451 4.4316425 16470 2292 4.6430583 33523 1180 5.110788 33523 779 4.924078 51585 746 4.4878364 15029 4092 4.779641 18020 6228 3.8732393 30472 5376 4.4120226 1848 2840 5.4418874 24952 2638 5.2496567 36558 2627 5.199316 40943 5640 3.6769154 10668 7401 5.2865014 29957 4497 3.9111974 48698 8059 2.9496775 13921 862 4.7236485 39174 4760 3.9191642 30498 3628 5.354149 30557 3248 5.43848 33997 9039 3.394697 40773 9032 4.6396866 46186 562 5.0802975 21425 7401 4.5135655 50609 562 4.3230453 50609 464 4.3052945 7877 3093 4.696365 9189 4868 3.0412648 14861 1905 4.640318 25932 9222 4.870158 34893 7131 4.346138 49951 9055 5.0066576 3857 1597 4.1393723 5309 562 4.790769 10664 6400 6.084344 20709 1657 4.6188126 35509 1499 4.9694543 7361 1338 4.794663 29745 7831 4.423161 17623 562 4.3465834 29758 1180 5.2529044 38444 7131 5.5920863 40577 192 5.5498123 3927 7401 4.726795 17124 2060 5.5406966 45806 1308 5.637669 46416 39 4.1354246 2781 5670 5.122589 3819 2627 4.463003 39527 3248 3.074817 45712 3124 4.786623 2786 2590 3.5974693 12671 1810 3.060909 32236 562 5.1731644 34709 3825 5.3521485 38252 2292 5.35800

It is interesting to see that some predictions made by ALS are outside the scale of 1 to 5, the rating scale in the original data set. I searched around and found that this has to do with ALS's approximation methods and it makes more sense to use the predictions as rankings for recommendations rather than simple predicted rating scores. A higher predicted rating score implies that the user is more interested in this item than that of a lower rating.

In [51]:
%sql
-- select minimum, maximum and average of prediction in to read list
SELECT MIN(b.prediction), MAX(b.prediction), AVG(b.prediction)
FROM to_read a
INNER JOIN tidy_rec b
ON a.user_id = b.user_id AND a.book_id = b.book_id;

min(prediction) max(prediction) avg(prediction) 0.96303517 6.084344 4.591680483160807

We can see that the average predicted rating in the to read list is 4.59, which is high and reasonable given that this is a reader's to do list and books in this list are those a reader has interests in. The minimum rating 0.96 is not necessarily a mistake or poor performance of the model, I believe there are circumstances that a reader may want to try something different but the model predicts that he/she will not like it or he/she will give a low rating score to that.